### PR TITLE
CubedSphere grid construction + tests

### PR DESCRIPTION
**Pull request by navidcy**
The cubed sphere grid from the MIT as read from DataDeps has the faces ordered:
  1. `+ x` direction
  2. `+ y` direction
  3. `+ z` direction (Arctic)
  4. `- x` direction
  5. `- y` direction
  6. `- z` direction (Antarctica)
 
So to avoid confusion for (users and developers!) I thought I make sure that the ConformalCubedSphere follows the same convention.

This PR does that + it adds tests to check whether the ConformalCubedSphereGrid metrics from the grid read from file and the grid constructed in Oceananigans.jl are the same.

After this is merged we'll move on to #2867 to constructing `ConformalCubedSphereGrid` using MultiRegion (an possibly nuke the `CubedSpheres` module completely). We will have a `ConformalCubedSphereGrid` (which is the same as the `cs32` grid read from file) to bench our MultiRegion implementation with!

(This PR also moves some of the spherical trigonometry utility function to `grid_utils.jl`.)

### CODE REVIEW

**Comment by navidcy** [comment]
this is ready for review :)

File: src/Grids/grid_utils.jl [hunk 1]
```diff
@@ -431,3 +432,101 @@ coordinate_summary(Δ::Union{AbstractVector, AbstractMatrix}, name) =
...
+Return the area of a spherical triangle on the unit sphere with sides `a`, `b`, and `c`.
+
+The area of a spherical triangle on the unit sphere is ``E = A + B + C - π``, where ``A``, ``B``, and ``C``
+are the triangle's inner angles.
+
+It has been known since Euler and Lagrange that ``\\tan(E/2) = P / (1 + \\cos a + \\cos b + \\cos c)``, where
+``P = (1 - \\cos²a - \\cos²b - \\cos²c + 2 \\cos a \\cos b \\cos c)^{1/2}``.
+"""
+function spherical_area_triangle(a::Number, b::Number, c::Number)
+    cosa, cosb, cosc = cos.((a, b, c))
```
Comment (glwagner) [inline src/Grids/grid_utils.jl:452]:
```suggestion
    cosa = cos(a)
    cosb = cos(b)
    cosc = cos(c)
```

File: src/Grids/orthogonal_spherical_shell_grid.jl [hunk 2]
```diff
@@ -127,7 +133,7 @@ function OrthogonalSphericalShellGrid(architecture::AbstractArchitecture = CPU()
                                       rotation = nothing)
 
     @warn "OrthogonalSphericalShellGrid is still under development. Use with caution!"
-    @warn "Azᶜᶜᵃ are correct for interior points, but Azᶠᶜᵃ, Azᶜᶠᵃ, Azᶠᶠᵃ are wrong close to boundaries"
+    @warn "Metrics close to boundaries might be wrong!"
```
Comment (glwagner) [inline src/Grids/orthogonal_spherical_shell_grid.jl:136]: why?

File: src/CubedSpheres/conformal_cubed_sphere_grid.jl [hunk 3]
```diff
@@ -127,28 +127,28 @@ function ConformalCubedSphereGrid(arch = CPU(), FT=Float64; face_size, z, face_h
 
     size, halo = face_size, face_halo
 
-    # +z face (face 1)
-    z⁺_face_grid = OrthogonalSphericalShellGrid(arch, FT; size, z, halo, radius, rotation=nothing)
+    # +x face (face 1)
```
Comment (glwagner) [inline src/CubedSpheres/conformal_cubed_sphere_grid.jl:130]: We need a sketch with lat-lon lines overlaying the cube so we understand the "+z", etc terminology!

File: src/Grids/orthogonal_spherical_shell_grid.jl [hunk 4]
```diff
@@ -193,90 +203,112 @@ function OrthogonalSphericalShellGrid(architecture::AbstractArchitecture = CPU()
...
 
-    ΔS = [Δxᶜᶜᵃ,  Δxᶠᶜᵃ,  Δxᶜᶠᵃ,  Δxᶠᶠᵃ,  Δyᶜᶜᵃ,  Δyᶜᶠᵃ,  Δyᶠᶜᵃ,  Δyᶠᶠᵃ]
+    ΔS = [Δxᶜᶜᵃ,  Δxᶠᶜᵃ,  Δxᶜᶠᵃ,  Δxᶠᶠᵃ,  Δyᶜᶜᵃ,  Δyᶜᶠᵃ,  Δyᶠᶜᵃ,  Δyᶠᶠᵃ ]
     ΔΣ = [Δσxᶜᶜᵃ, Δσxᶠᶜᵃ, Δσxᶜᶠᵃ, Δσxᶠᶠᵃ, Δσyᶜᶜᵃ, Δσyᶜᶠᵃ, Δσyᶠᶜᵃ, Δσyᶠᶠᵃ]
 
-    # convert degrees to meters
     for (Δs, Δσ) in zip(ΔS, ΔΣ)
         @. Δσ = deg2rad(Δσ)
         @. Δs = radius * Δσ
     end
```
Comment (glwagner) [inline src/Grids/orthogonal_spherical_shell_grid.jl:294]: It's a bit hard to read like this, would it be better to simply write explicitly things like

```julia
@. Δxᶜᶜᵃ = radius * Δσxᶜᶜᵃ
```

etc?

Comment (glwagner) [reply src/Grids/orthogonal_spherical_shell_grid.jl:294]: I think I prefer not using the loop

**Review by glwagner — APPROVED** [review]

File: src/CubedSpheres/conformal_cubed_sphere_grid.jl (hunk 3 above)
Comment (navidcy) [reply src/CubedSpheres/conformal_cubed_sphere_grid.jl:130]: Yeap, I plan to add some figures in the docs. For now, is the figures at https://clima.github.io/CubedSphere.jl/stable/conformal_cubed_sphere/ helpful? +z is the face that looks in the positive z direction.

File: src/Grids/grid_utils.jl [hunk 5]
```diff
@@ -431,3 +432,101 @@ coordinate_summary(Δ::Union{AbstractVector, AbstractMatrix}, name) =
...
+"""
+    spherical_area_triangle(a::Number, b::Number, c::Number)
+
+Return the area of a spherical triangle on the unit sphere with sides `a`, `b`, and `c`.
+
+The area of a spherical triangle on the unit sphere is ``E = A + B + C - π``, where ``A``, ``B``, and ``C``
+are the triangle's inner angles.
+
+It has been known since Euler and Lagrange that ``\\tan(E/2) = P / (1 + \\cos a + \\cos b + \\cos c)``, where
+``P = (1 - \\cos²a - \\cos²b - \\cos²c + 2 \\cos a \\cos b \\cos c)^{1/2}``.
```
[the remark this answers is not in the document]
Comment (navidcy) [reply src/Grids/grid_utils.jl:449]: I added citations to the works by Euler and Lagrange. ;)

**Comment by navidcy** [comment]
@glwagner asks why we have the warning saying that the metrics near the edge of the `OrthogonalSphericalShellGrid` might be wrong. I should have clarified there that I "The _area_ metrics might be wrong near the boundaries".

I explain here:

To compute the areas near the edge we actually need to know the node locations of the halos, and to do that we need to implement proper halo filling. Why is that? Take for For example, consider a face near the edge, then:

```julia
Δxᶠᶠᵃ[1, j, k] = xᶜᶠᵃ[1, j, k] - xᶜᶠᵃ[0, j, k]
```

but we can't compute the lat-lon position of `xᶜᶠᵃ[0, j, k]` from the conformal map because it belongs to a different panel. However, from symmetry argument, one can say that this would be just twice the distance from the first cell center to the edge of the panel:

```julia
Δxᶠᶠᵃ[1, j, k] = 2 * (xᶜᶠᵃ[1, j, k] - xᶠᶠᵃ[1, j, k])
```

So that's that. The distance metrics are actually correct, _even near the edges_.

How do we compute now areas? An approximation is 

```Julia
    @. Azᶜᶜᵃ = Δxᶜᶜᵃ * Δyᶜᶜᵃ
    @. Azᶠᶜᵃ = Δxᶠᶜᵃ * Δyᶠᶜᵃ
    @. Azᶜᶠᵃ = Δxᶜᶠᵃ * Δyᶜᶠᵃ
    @. Azᶠᶠᵃ = Δxᶠᶠᵃ * Δyᶠᶠᵃ
```

But that's only an approximation because these quadrilaterals are not symmetrical. Anywhere in the interior of the grid we can improve on this by computing the exact area of the spherical quadrilateral given that we know the location of each of its vertices. But for a quadrilateral that touches the edge some of its vertices would belong to the next panels, i.e., to get their location we'd need proper halo filling. E.g., `Azᶠᶜᵃ[i, j]` has vertices\

```julia
   (φᶜᶠᵃ[i-1,  j ], λᶠᶠᵃ[i-1,  j ])
   (φᶜᶠᵃ[ i ,  j ], λᶠᶠᵃ[ i ,  j ])
   (φᶜᶠᵃ[ i , j+1], λᶠᶠᵃ[ i , j+1])
   (φᶜᶠᵃ[i-1, j+1], λᶠᶠᵃ[i-1, j+1])
```

thus `Azᶠᶜᵃ[1, j]` would require the location of `(φᶜᶠᵃ[0, j+1], λᶠᶠᵃ[0, j+1])` and `(φᶜᶠᵃ[0,  j ], λᶠᶠᵃ[0,  j ])`.

But now that I'm thinking about it, again from symmetry we would expect that the area of quadrilateral with vertices at

```julia
   (φᶜᶠᵃ[ 0 ,  j ], λᶜᶠᵃ[ 0 ,  j ])
   (φᶜᶠᵃ[ 1 ,  j ], λᶜᶠᵃ[ 1 ,  j ])
   (φᶜᶠᵃ[ 1 , j+1], λᶜᶠᵃ[ 1 , j+1])
   (φᶜᶠᵃ[ 0 , j+1], λᶜᶠᵃ[ 0 , j+1])
```

should be double the area of the quadrilateral with vertices

```julia
   (φᶠᶠᵃ[ 1 ,  j ], λᶠᶠᵃ[ 1 ,  j ])
   (φᶜᶠᵃ[ 1 ,  j ], λᶜᶠᵃ[ 1 ,  j ])
   (φᶜᶠᵃ[ 1 , j+1], λᶜᶠᵃ[ 1 , j+1])
   (φᶠᶠᵃ[ 1 , j+1], λᶠᶠᵃ[ 1 , j+1])
```

I'll fix this and remove the warning!

File: src/Grids/orthogonal_spherical_shell_grid.jl (hunk 4 above)
Comment (navidcy) [reply src/Grids/orthogonal_spherical_shell_grid.jl:294]: done

File: src/Grids/grid_utils.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (simone-silvestri) [reply src/Grids/grid_utils.jl:449]: Ahah nice 18th century references.

**Pull request merged**